[AIRFLOW-3918] Add ssh private-key support to git-sync for KubernetesExecutor

dmateusp opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

> - [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
>  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
>  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

https://issues.apache.org/jira/browse/AIRFLOW-3918

### Description

> - [X] Here are some details about my PR, including screenshots of any UI changes:

This PR adds support for Git Sync authentication through SSH key (e.g. a GitHub deployment read-only key)

### Tests

> - [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

* test_worker_configuration_auth_both_ssh_and_user: tests that configuration validation throws an exception when both user authentication and SSH authentication are configured
* test_init_environment_using_git_sync_ssh: tests that the environment is correctly set up for the init pod when using SSH auth
* test_make_pod_git_sync_ssh: tests that the make_pod function correctly generates the config for the pod, covers security_context test

Additionally tested at HBC on our cluster

### Commits

> - [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
>  1. Subject is separated from body by a blank line
>  1. Subject is limited to 50 characters (not including Jira issue reference)
>  1. Subject does not end with a period
>  1. Subject uses the imperative mood ("add", not "adding")
>  1. Body wraps at 72 characters
>  1. Body explains "what" and "why", not "how"

### Documentation

> - [X] In case of new functionality, my PR adds documentation that describes how to use it.
>  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
>  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

> - [X] Passes `flake8`

dmateusp committed:
Add configuration for git SSH auth and update git-sync version in
template

Change default airflow config to have keys for SSH authentication
(mutually exclusive of user authentication)

Update git-sync version to the latest, current version did not
support SSH authentication environment variables

dmateusp committed:
Add tests for git-sync SSH authentication

Tests that the resulting configs are valid and that security_context
is correctly added when using make_pod

dmateusp committed:
Add security_context support to make_pod

Security context was required to read the mounted SSH key for
git-sync SSH authentication

dmateusp committed:
Parse SSH git-sync authentication keys

ashb commented on airflow/config_templates/default_airflow.cfg:
```diff
@@ -640,18 +640,23 @@ logs_volume_host =
 git_dags_folder_mount_point =
+# Kubernetes secret name and secret key where the SSH key is stored
+git_ssh_key_secret_name =
+git_ssh_key_secret_key =
```
What is the key_secret_key? I can't work this out from having no context in to git-sync :)

dmateusp replied:
In a Kubernetes secret definition:
```
apiVersion: v1
kind: Secret
metadata:
  name: airflow-secrets
data:
  gitSshKey: <base64encodeddata>

git_ssh_key_secret_name -> airflow-secrets
git_ssh_key_secret_key    -> gitSshKey
```

Should I add this as a comment ?

dmateusp replied:
Final version could be:

```# Git-sync SSH authentication
# i.e. Kubernetes secrets:
# apiVersion: v1
# kind: Secret
# metadata:
#   name: airflow-secrets
# data:
#   gitSshKey: <base64_encoded_data>
#   knownHosts: <base64_encoded_data>
#
# git_ssh_key_secret_name = airflow-secrets
# git_ssh_key_secret_key  = gitSshKey
# git_ssh_known_hosts_secret_name = airflow-secrets
# git_ssh_known_hosts_secret_key = knownHosts
# Kubernetes secret name and secret key where the SSH key is stored
git_ssh_key_secret_name =
git_ssh_key_secret_key =
# Kubernetes secret name and secret key where the SSH known hosts file data is stored
git_ssh_known_hosts_secret_name =
git_ssh_known_hosts_secret_key =

ashb replied:
Ah the "dict key" inside the secret. Does it actually need to be configurable, or could we just say "the secret must have look like "this"?

known_hosts could probably be a config map, not a secret (as it contains public keys, not private). Again - do we need the `*_key` suffix for known hosts either?

dmateusp replied:
I can enforce some specific key names, and point taken on known_hosts I'll change to a config map now

dmateusp replied:
I could reduce it to 1 config for each, by using the same syntax as in [kubernetes_secrets]

`secret-or-configmap-name=key-in-config`

dmateusp replied:
```
# To get Git-sync SSH authentication set up follow this format
#
# airflow-secrets.yaml:
# ---
# apiVersion: v1
# kind: Secret
# metadata:
#   name: airflow-secrets
# data:
#   gitSshKey: <base64_encoded_data>
# ---
# airflow-configmap.yaml:
# apiVersion: v1
# kind: ConfigMap
# metadata:
#   name: airflow-configmap
# data:
#   knownHosts: |
#       github.com ssh-rsa <...>
#
# git_ssh_secret_name = airflow-secrets
# git_ssh_known_hosts_configmap_name = airflow-configmap
git_ssh_secret_name =
git_ssh_known_hosts_configmap_name =
```

I'll implement this, it keeps the airflow.cfg options smaller

ashb commented on airflow/contrib/kubernetes/worker_configuration.py:
```diff
@@ -78,19 +77,37 @@ def _get_init_containers(self, volume_mounts):
+                },
+                {
+                    'name': 'GIT_KNOWN_HOSTS',
+                    'value': 'false'
```
This should default to true - having false hard-coded is a security no-no.

dmateusp replied:
Got it, should I add the know_hosts the same way I added the SSH key option ?

ashb commented on airflow/contrib/kubernetes/worker_configuration.py:
```diff
@@ -213,8 +253,7 @@ def generate_dag_volume_mount_path(self):
         volumes_dict, volume_mounts_dict = self.init_volumes_and_mounts()
-        worker_init_container_spec = self._get_init_containers(
-            copy.deepcopy(volume_mounts_dict))
+        worker_init_container_spec = self._get_init_containers()
```
Why does this ignore the volumes_dict returned form `init_volumes_and_mounts`?

dmateusp replied:
There are three volumes: the dags volume, the logs volume and the git ssh secret volume (if SSH is turned on)

with the deep copy you had the init container having all of the mounts that the pod container has, but there is no logical need for the init container:
- to share the same mount path for the dags (and we can manage security for the init pod better)
 - no needs for the Airflow logs to be mounted
 - the git SSH secret can be mounted to the init container without it being mounted on the worker pods (so you only have it on the worker during initialization, no need for SSH github secret in the worker post initialization since the DAGs are pulled once)

ashb replied:
Ahhhh gotcha. Nice

dmateusp committed:
Reduce git SSH authentication config parameters to 2

Hardcode Kubernetes secret key name and known hosts configmap key name

Add example of configmap + Kubernetes secret snippet in config template

dmateusp committed:
Rename git_ssh_secret_name to git_ssh_key_secret_name

ashb commented on airflow/contrib/kubernetes/worker_configuration.py:
```diff
@@ -213,8 +281,7 @@ def generate_dag_volume_mount_path(self):
     def make_pod(self, namespace, worker_uuid, pod_id, dag_id, task_id, execution_date,
                  try_number, airflow_command, kube_executor_config):
         volumes_dict, volume_mounts_dict = self.init_volumes_and_mounts()
```
Can we rename this? The `init_` prefix on it is confusing me with `init_containers` - and I think now these volumes/mounts have nothing to do with the init_container, correct?

dmateusp replied:
My understanding is that it never was for the `init` container, it used to pass the return value to the init container function (see below), I think `init_volumes_and_mounts` was meant for "initialize volumes and mounts", I agree it's confusing but my code change doesn't make it different than it was

I'll rename for clarity

ashb replied:
Yup, your PR doesn't change it, but at least before it was also used by the init container, where as now it isn't :)

Thanks.

ashb commented on tests/contrib/executors/test_kubernetes_executor.py:
```diff
@@ -161,6 +165,39 @@ def test_worker_configuration_no_subpaths(self):
+        mock_conf_get.side_effect = get_conf
+        mock_config_as_dict.return_value = {'core': ''}
+        with self.assertRaises(AirflowConfigException):
+            KubeConfig()
```
What error are we expecting here? This test feels like it would be hard to debug if it breaks in the future. Either a check on the message, and/or a comment saying _why_ we expect this to raise please.

dmateusp replied:
is it better?

ashb replied:
Thanks.

ashb commented on tests/contrib/executors/test_kubernetes_executor.py:
```diff
@@ -236,6 +273,103 @@ def test_worker_environment_dags_folder_using_git_sync(self):
+        init_containers = worker_config._get_init_containers()
+        git_ssh_known_hosts_file = next((x['value'] for x in init_containers[0]['env']
+                                         if x['name'] == 'GIT_SSH_KNOWN_HOSTS_FILE'), None)
+        print(init_containers[0]['volumeMounts'])
```
Remove please :)
```suggestion
```

ashb commented on tests/contrib/executors/test_kubernetes_executor.py:
```diff
@@ -236,6 +273,103 @@ def test_worker_environment_dags_folder_using_git_sync(self):
+        self.kube_config.dags_volume_host = None
+        self.kube_config.dags_in_image = None
+
+        worker_config = WorkerConfiguration(self.kube_config)
```
A test that `worker_config.git_sync_ssh_known_hosts_volume_name` is set to something would be good - we are using it in a later test, but we are never checking that this is set to anything directly

dmateusp replied:
`git_sync_ssh_known_hosts_volume_name` is a property set in the Class at initialization, (it does not get it's value from config but rather a hardcoded value)

I can add the following test that checks that those hardcoded values are set but I'm not sure of the value

```python
    def test_worker_config_properties(self):
        worker_config = WorkerConfiguration(self.kube_config)

        self.assertTrue(worker_config.dags_volume_name)
        self.assertTrue(worker_config.logs_volume_name)
        self.assertTrue(worker_config.git_sync_ssh_secret_volume_name)
        self.assertTrue(worker_config.git_ssh_key_secret_key)
        self.assertTrue(worker_config.git_sync_ssh_known_hosts_volume_name)
        self.assertTrue(worker_config.git_ssh_known_hosts_configmap_key)
```

ashb replied:
Oh yes, I hadn't twigged that. This is fine then, (though perhaps changing it `worker_config.GIT_SYNC_SSH_KNOWN_HOSTS_VOLUME_NAME` etc. That's a bit ... shouty ... though. Probably leave it as is.)

ashb commented on tests/contrib/executors/test_kubernetes_executor.py:
```diff
@@ -236,6 +273,103 @@ def test_worker_environment_dags_folder_using_git_sync(self):
+        self.assertTrue(volume_mount_ssh_known_hosts_file)
+        self.assertEqual(git_ssh_known_hosts_file,
+                         volume_mount_ssh_known_hosts_file,
+                         ('The location where the git known hosts file is mounted'
```
Style: remove extra brackets, I don't think they are needed by PEP8

```suggestion
                         'The location where the git known hosts file is mounted'
```

ashb commented on airflow/contrib/kubernetes/worker_configuration.py:
```diff
@@ -183,6 +229,28 @@ def _construct_volume(name, claim, host):
+                    }]
+                }
+            }
+            if self.kube_config.git_ssh_known_hosts_configmap_name:
```
The known_hosts confimap should probably work even with user+password SSH 
```suggestion
        if self.kube_config.git_ssh_known_hosts_configmap_name:
```

ashb commented on airflow/contrib/kubernetes/worker_configuration.py:
```diff
@@ -183,6 +229,28 @@ def _construct_volume(name, claim, host):
+                    'items': [{
+                        'key': self.git_ssh_key_secret_key,
+                        'path': 'ssh',
+                        'mode': 288
```
Little bit easier to understand the mode if we use octal:

```suggestion
                        'mode': 0o440
```

ashb commented on airflow/contrib/kubernetes/worker_configuration.py:
```diff
@@ -183,6 +229,28 @@ def _construct_volume(name, claim, host):
+                    'configMap': {
+                        'name': self.kube_config.git_ssh_known_hosts_configmap_name
+                    },
+                    'mode': 288
```
```suggestion
                    'mode': 0o440
```

ashb reviewed (changes requested):
Few minor changes - some of the "suggestions" need more than just the one line requested, but github doesn't allow multi line suggestions.

Looking good. Please ping me again once you've done the changes (in Slack is better, my inbox is a _mess_ right now)

ashb commented on airflow/contrib/kubernetes/worker_configuration.py:
```diff
@@ -37,10 +35,14 @@ def __init__(self, kube_config):
+        self.git_sync_ssh_secret_volume_name = 'git-sync-ssh-key'
+        self.git_ssh_key_secret_key = 'gitSshKey'
+        self.git_sync_ssh_known_hosts_volume_name = 'git-sync-known-hosts'
+        self.git_ssh_known_hosts_configmap_key = 'known_hosts'
```
Oh, since these are hard-coded lets move them to "class properties":

```python
class WorkerConfiguration(LoggingMixin):
    """Contains Kubernetes Airflow Worker configuration logic"""

    dags_volume_name = 'airflow-dags'
    logs_volume_name = 'airflow-logs'
    git_sync_ssh_secret_volume_name = 'git-sync-ssh-key'
    git_ssh_key_secret_key = 'gitSshKey'
    git_sync_ssh_known_hosts_volume_name = 'git-sync-known-hosts'
    git_ssh_known_hosts_configmap_key = 'known_hosts'

    def __init__(self, kube_config):
```

Nothing else needs to change about how these are used.

dmateusp committed:
Incorporate PR feedback

Move hardcoded Class instance attributes to Class properties

Use octal to represent file permissions

Move known hosts configuration to allow usage with username/password

Make ConfigurationException test more specific

Remove print and extra parenthesis

Rename init_volumes_and_mounts to _get_volumes_and_mounts for
clarity

Add tests for username/password and known_hosts configuration
